// @mintlify/validation/dist/editor-navigation/order.js
import { generateKeyBetween, generateNKeysBetween } from 'fractional-indexing';
export { generateKeyBetween, generateNKeysBetween };
export function compareOrder(a, b) {
    if (typeof a === 'number' && typeof b === 'number')
        return a - b;
    if (typeof a === 'number')
        return -1;
    if (typeof b === 'number')
        return 1;
    return a < b ? -1 : a > b ? 1 : 0;
}
export function compareNodes(a, b) {
    const byOrder = compareOrder(a.order, b.order);
    if (byOrder !== 0)
        return byOrder;
    return a.id < b.id ? -1 : a.id > b.id ? 1 : 0;
}
export function sortByOrder(nodes) {
    return [...nodes].sort(compareNodes);
}
function groupByParent(nodes) {
    var _a;
    const byParent = new Map();
    for (const node of nodes) {
        const group = (_a = byParent.get(node.parent)) !== null && _a !== void 0 ? _a : [];
        group.push(node);
        byParent.set(node.parent, group);
    }
    return byParent;
}
export function groupSiblingsByParent(nodes) {
    const byParent = groupByParent(nodes);
    for (const [parent, group] of byParent) {
        byParent.set(parent, sortByOrder(group));
    }
    return byParent;
}
function applyOrderUpdates(tree, orderById) {
    if (orderById.size === 0)
        return tree;
    return tree.map((node) => {
        const order = orderById.get(node.id);
        return order === undefined ? node : Object.assign(Object.assign({}, node), { order });
    });
}
export function normalizeTreeOrder(tree) {
    const orderById = new Map();
    for (const group of groupByParent(tree).values()) {
        if (!group.some((node) => typeof node.order === 'number'))
            continue;
        const sorted = sortByOrder(group);
        const keys = generateNKeysBetween(null, null, sorted.length);
        sorted.forEach((node, index) => {
            const key = keys[index];
            if (key !== undefined && node.order !== key)
                orderById.set(node.id, key);
        });
    }
    return applyOrderUpdates(tree, orderById);
}
export function alignTreeOrderKeys(tree, reference) {
    var _a, _b, _c;
    const referenceKeys = new Map();
    for (const node of reference) {
        if (typeof node.order !== 'string')
            continue;
        referenceKeys.set(`${node.parent}::${node.id}`, node.order);
    }
    const orderById = new Map();
    for (const group of groupByParent(tree).values()) {
        const sorted = sortByOrder(group);
        const keys = sorted.map(() => null);
        let lastReused = null;
        sorted.forEach((node, index) => {
            const key = referenceKeys.get(`${node.parent}::${node.id}`);
            if (key !== undefined && (lastReused === null || key > lastReused)) {
                keys[index] = key;
                lastReused = key;
            }
        });
        let start = 0;
        while (start < sorted.length) {
            if (keys[start] !== null) {
                start++;
                continue;
            }
            let end = start;
            while (end < sorted.length && keys[end] === null)
                end++;
            const before = start > 0 ? ((_a = keys[start - 1]) !== null && _a !== void 0 ? _a : null) : null;
            const after = end < sorted.length ? ((_b = keys[end]) !== null && _b !== void 0 ? _b : null) : null;
            const filled = generateNKeysBetween(before, after, end - start);
            for (let i = start; i < end; i++)
                keys[i] = (_c = filled[i - start]) !== null && _c !== void 0 ? _c : null;
            start = end;
        }
        sorted.forEach((node, index) => {
            var _a;
            const key = (_a = keys[index]) !== null && _a !== void 0 ? _a : null;
            if (key !== null && node.order !== key)
                orderById.set(node.id, key);
        });
    }
    return applyOrderUpdates(tree, orderById);
}
