// @mintlify/validation/dist/mint-config/validateConfig.js
import { z } from 'zod';
import { customErrorMap } from './customErrorMap.js';
import { refineMissingVersions } from './refinements/refineMissingVersions.js';
import { mintConfigSchema } from './schemas/v1/config.js';
import { docsConfigSchema } from './schemas/v2/index.js';
import { sourceRefSchema, } from './schemas/v2/properties/navigation/sourceRef.js';
import { aggregateWarnings } from './warnings/aggregateWarnings.js';
const sourceRefArrayKeys = new Set([
    'anchors',
    'dropdowns',
    'languages',
    'menu',
    'pages',
    'products',
    'tabs',
    'versions',
    'groups',
]);
const sourceRefPlaceholders = {
    anchors: { anchor: 'sourceRef', href: '/sourceRef' },
    dropdowns: { dropdown: 'sourceRef', href: '/sourceRef' },
    languages: { language: 'en', href: '/sourceRef' },
    menu: { item: 'sourceRef', href: '/sourceRef' },
    pages: 'sourceRef',
    products: { product: 'sourceRef', href: '/sourceRef' },
    tabs: { tab: 'sourceRef', href: '/sourceRef' },
    versions: { version: 'sourceRef', href: '/sourceRef' },
    groups: { group: 'sourceRef', pages: [] },
};
const strictSourceRefSchema = sourceRefSchema.strict();
function hasSourceRefKey(value) {
    return (typeof value === 'object' && value !== null && !Array.isArray(value) && 'sourceRef' in value);
}
function isSourceRef(value) {
    return hasSourceRefKey(value) && typeof value.sourceRef === 'string';
}
function validateSourceRefNodes(value, parentKey, path = []) {
    if (hasSourceRefKey(value)) {
        if (!parentKey || !sourceRefArrayKeys.has(parentKey)) {
            return [
                {
                    code: z.ZodIssueCode.custom,
                    path,
                    message: 'sourceRef is only supported as an element in navigation arrays',
                },
            ];
        }
        const result = strictSourceRefSchema.safeParse(value);
        if (!result.success) {
            return result.error.issues.map((issue) => (Object.assign(Object.assign({}, issue), { path: [...path, ...issue.path] })));
        }
        return [];
    }
    if (Array.isArray(value)) {
        return value.flatMap((entry, index) => validateSourceRefNodes(entry, parentKey, [...path, index]));
    }
    if (typeof value !== 'object' || value === null) {
        return [];
    }
    return Object.entries(value).flatMap(([key, entry]) => validateSourceRefNodes(entry, key, [...path, key]));
}
function sanitizeSourceRefsForValidation(value, parentKey) {
    if (isSourceRef(value) && parentKey && sourceRefArrayKeys.has(parentKey)) {
        return sourceRefPlaceholders[parentKey];
    }
    if (Array.isArray(value)) {
        return value.map((entry) => sanitizeSourceRefsForValidation(entry, parentKey));
    }
    if (typeof value !== 'object' || value === null) {
        return value;
    }
    return Object.fromEntries(Object.entries(value).map(([key, entry]) => [key, sanitizeSourceRefsForValidation(entry, key)]));
}
export function validateMintConfig(value) {
    // save old error map, as this error map is configured globally, and we don't want to overwrite global behavior
    const oldErrorMap = z.getErrorMap();
    z.setErrorMap(customErrorMap);
    const refinedSchema = mintConfigSchema.superRefine(refineMissingVersions);
    const validationResults = refinedSchema.safeParse(value);
    // reset error map
    z.setErrorMap(oldErrorMap);
    if (!validationResults.success)
        return validationResults;
    return Object.assign(Object.assign({}, validationResults), { warnings: aggregateWarnings(validationResults.data) });
}
export function validateDocsConfig(value) {
    const oldErrorMap = z.getErrorMap();
    z.setErrorMap(customErrorMap);
    const validationResults = docsConfigSchema.safeParse(value);
    z.setErrorMap(oldErrorMap);
    if (!validationResults.success)
        return validationResults;
    return Object.assign(Object.assign({}, validationResults), { warnings: [] });
}
export function validateSourceRefDocsConfig(value) {
    const oldErrorMap = z.getErrorMap();
    z.setErrorMap(customErrorMap);
    const sourceRefIssues = validateSourceRefNodes(value);
    const validationResults = sourceRefIssues.length > 0
        ? {
            success: false,
            error: new z.ZodError(sourceRefIssues),
        }
        : docsConfigSchema.safeParse(sanitizeSourceRefsForValidation(value));
    z.setErrorMap(oldErrorMap);
    if (!validationResults.success)
        return validationResults;
    return Object.assign(Object.assign({}, validationResults), { data: value, warnings: [] });
}
