// @mintlify/prebuild/dist/prebuild/categorizeFilePaths.js
import { validate, validateAsyncApi, buildImportMap, getFileCategory, isSnippetExtension, getAST, } from '@mintlify/common';
import { readFile } from 'fs/promises';
import yaml from 'js-yaml';
import * as path from 'path';
import { formatError } from '../errorMessages/formatError.js';
import { getFileList } from '../fs/index.js';
import { getFileExtension } from '../utils.js';
export const categorizeFilePaths = async (contentDirectoryPath, mintIgnore = [], disableOpenApi) => {
    const allFiles = getFileList(contentDirectoryPath, contentDirectoryPath, mintIgnore);
    const mdxFiles = [];
    const nonMdxFiles = [];
    for await (const filename of allFiles) {
        const extension = getFileExtension(filename);
        if (isSnippetExtension(extension)) {
            const filePath = path.join(contentDirectoryPath, filename);
            const content = await readFile(filePath, 'utf8');
            try {
                const tree = getAST(content, filePath);
                mdxFiles.push({ path: filename, tree });
            }
            catch (error) {
                console.error(formatError(error, filePath, contentDirectoryPath));
            }
        }
        else {
            nonMdxFiles.push({ filename, extension });
        }
    }
    const { importedFiles, fileImportsMap } = buildImportMap(mdxFiles);
    const contentFilenames = [];
    const snippets = [];
    const snippetsV2 = [];
    for (const file of mdxFiles) {
        const category = getFileCategory(file.path, { importedFiles });
        switch (category) {
            case 'snippet':
                snippets.push(file.path);
                break;
            case 'snippet-v2':
                snippetsV2.push(file.path);
                break;
            case 'page':
                contentFilenames.push(file.path);
                break;
        }
    }
    const staticFilenames = [];
    const openApiFiles = [];
    const asyncApiFiles = [];
    const invalidSpecFiles = [];
    for (const { filename, extension } of nonMdxFiles) {
        switch (extension) {
            case 'json':
            case 'yaml':
            case 'yml':
                const filePath = path.join(contentDirectoryPath, filename);
                const str = await readFile(filePath, 'utf8');
                let obj;
                try {
                    obj = yaml.load(str);
                }
                catch (error) {
                    invalidSpecFiles.push({
                        originalFileLocation: filename,
                        reason: `Invalid JSON or YAML syntax: ${error instanceof Error ? error.message : String(error)}`,
                        kind: 'syntax',
                    });
                    break;
                }
                if (!obj || typeof obj !== 'object')
                    break;
                const isOpenApi = Object.keys(obj).includes('openapi');
                const isAsyncApi = Object.keys(obj).includes('asyncapi');
                const fileName = path.parse(filename).name;
                if (isOpenApi && !disableOpenApi) {
                    try {
                        const { schema: openApiDocument } = await validate(obj);
                        if (openApiDocument) {
                            openApiFiles.push({
                                filename: fileName,
                                spec: obj,
                                originalFileLocation: filename,
                            });
                        }
                    }
                    catch (error) {
                        invalidSpecFiles.push({
                            originalFileLocation: filename,
                            reason: error instanceof Error ? error.message : String(error),
                            kind: 'openapi',
                        });
                    }
                }
                if (isAsyncApi) {
                    try {
                        const { document: asyncApiDocument } = await validateAsyncApi(str);
                        if (asyncApiDocument) {
                            asyncApiFiles.push({
                                filename: fileName,
                                spec: asyncApiDocument,
                                originalFileLocation: filename,
                            });
                        }
                    }
                    catch (error) {
                        invalidSpecFiles.push({
                            originalFileLocation: filename,
                            reason: error instanceof Error ? error.message : String(error),
                            kind: 'asyncapi',
                        });
                    }
                }
                break;
            default:
                staticFilenames.push(filename);
        }
    }
    return {
        contentFilenames,
        staticFilenames,
        openApiFiles,
        asyncApiFiles,
        invalidSpecFiles,
        snippets,
        snippetsV2,
        fileImportsMap,
    };
};
