// @mintlify/previewing/dist/local-preview/listener/docs-config-cache.js
let docsConfigVariablesHash;
let docsConfigApiHash;
let docsConfigNavigationHash;
const getJsonHash = (value) => {
    return JSON.stringify(value ?? {});
};
export const getVariablesHash = (variables) => {
    return getJsonHash(variables);
};
export const getDocsConfigApiHash = (docsConfig) => {
    return getJsonHash(docsConfig.api);
};
export const getDocsConfigNavigationHash = (docsConfig) => {
    return getJsonHash(docsConfig.navigation);
};
export const updateDocsConfigHashCache = (docsConfig) => {
    docsConfigVariablesHash = getVariablesHash(docsConfig.variables);
    docsConfigApiHash = getDocsConfigApiHash(docsConfig);
    docsConfigNavigationHash = getDocsConfigNavigationHash(docsConfig);
};
export const hasNavigationApiReference = (value) => {
    if (value == null || typeof value !== 'object') {
        return false;
    }
    if (Object.prototype.hasOwnProperty.call(value, 'openapi') ||
        Object.prototype.hasOwnProperty.call(value, 'asyncapi')) {
        return true;
    }
    return Object.values(value).some(hasNavigationApiReference);
};
export const hasDocsConfigApiChanged = (newApiHash) => {
    return docsConfigApiHash == null || docsConfigApiHash !== newApiHash;
};
export const hasDocsConfigNavigationChanged = (newNavigationHash) => {
    return docsConfigNavigationHash == null || docsConfigNavigationHash !== newNavigationHash;
};
export const hasDocsConfigVariablesChanged = (newVariablesHash) => {
    return docsConfigVariablesHash == null || docsConfigVariablesHash !== newVariablesHash;
};
export const clearDocsConfigCaches = () => {
    docsConfigVariablesHash = undefined;
    docsConfigApiHash = undefined;
    docsConfigNavigationHash = undefined;
};
