// @mintlify/previewing/dist/local-preview/listener/index.js
import { jsx as _jsx } from "react/jsx-runtime";
import { findAndRemoveImports, hasImports, getFileCategory, validate, stringifyTree, isMintIgnored, replaceVariables, } from '@mintlify/common';
import { createPage, getFileListSync, MintConfigUpdater, DocsConfigUpdater, preparseMdxTree, prebuild, } from '@mintlify/prebuild';
import Chalk from 'chalk';
import chokidar from 'chokidar';
import { promises as _promises } from 'fs';
import fse from 'fs-extra';
import fs from 'fs/promises';
import yaml from 'js-yaml';
import pathUtil from 'path';
import { CMD_EXEC_PATH, NEXT_PROPS_PATH, NEXT_PUBLIC_PATH, CLIENT_PATH } from '../../constants.js';
import { addChangeLog, addErrorLog, clearErrorLogs, getCurrentErrorLogs, } from '../../logging-state.js';
import { AddedLog, DeletedLog, EditedLog, WarningLog, InfoLog, ErrorLog } from '../../logs.js';
import { clearDocsConfigCaches, getDocsConfigApiHash, getDocsConfigNavigationHash, getVariablesHash, hasDocsConfigApiChanged, hasDocsConfigNavigationChanged, hasNavigationApiReference, hasDocsConfigVariablesChanged, updateDocsConfigHashCache, } from './docs-config-cache.js';
import { generateDependentSnippets } from './generateDependentSnippets.js';
import { generatePagesWithImports } from './generatePagesWithImports.js';
import { getDocsState } from './getDocsState.js';
import { initializeImportCache, updateImportCacheForFile, removeFromImportCache, getImportedFilesFromCache, syncImportedFileLocations, } from './importCache.js';
import { clearPageMetadataCache, removePageMetadataCacheEntryForFile, upsertPageMetadataCacheEntry, upsertPageMetadataCacheEntryForFile, } from './page-metadata-cache.js';
import { hasTrackedReferencedFile, refreshTrackedReferencedFiles } from './referencedFiles.js';
import { regenerateAllSnippets } from './regenerateAllSnippets.js';
import { resolveImportsFromImportClosure } from './resolve-imports-from-import-closure.js';
import { resolvePageImports } from './resolve-page-imports.js';
import { updateCustomLanguages, updateGeneratedNav, updateGeneratedNavFromPageMetadataCache, updateOpenApiFiles, upsertOpenApiFile, } from './update.js';
import { getCurrentVariables, getMintIgnoreGlobs, handleParseError, isFileSizeValid, isJsonValid, readJsonFile, seedFrontmatterHashForPage, shouldRegenerateNavForPage, suppressParseError, } from './utils.js';
const { readFile } = _promises;
const frontmatterHashes = new Map();
const SLOW_UPDATE_LOG_THRESHOLD_MS = 1000;
const getCurrentDocsConfig = async () => {
    const propsDocsJsonPath = pathUtil.join(NEXT_PROPS_PATH, 'docs.json');
    try {
        const propsDocsConfig = await readJsonFile(propsDocsJsonPath);
        return propsDocsConfig;
    }
    catch { }
    const docsJsonPath = pathUtil.join(CMD_EXEC_PATH, 'docs.json');
    return await DocsConfigUpdater.getConfig(docsJsonPath, false, CMD_EXEC_PATH, suppressParseError);
};
const writeDocsConfigWithCurrentNavigation = async (docsConfig) => {
    const propsDocsJsonPath = pathUtil.join(NEXT_PROPS_PATH, 'docs.json');
    try {
        const currentDocsConfig = await readJsonFile(propsDocsJsonPath);
        await DocsConfigUpdater.writeConfigFile({ ...docsConfig, navigation: currentDocsConfig.navigation }, CLIENT_PATH);
        return;
    }
    catch { }
    await DocsConfigUpdater.writeConfigFile(docsConfig, CLIENT_PATH);
};
const getCurrentRawDocsConfig = async () => {
    const docsJsonPath = pathUtil.join(CMD_EXEC_PATH, 'docs.json');
    if (!(await fse.pathExists(docsJsonPath))) {
        return undefined;
    }
    try {
        return await DocsConfigUpdater.getConfig(docsJsonPath, false, CMD_EXEC_PATH, suppressParseError);
    }
    catch {
        return undefined;
    }
};
const hasDocsConfigApiRoutes = (docsConfig) => {
    return docsConfig.api != null || hasNavigationApiReference(docsConfig.navigation);
};
const getElapsedMs = (startTime) => {
    return Number((process.hrtime.bigint() - startTime) / 1000000n);
};
const formatElapsedSeconds = (durationMs) => {
    return `${(durationMs / 1000).toFixed(2)}s`;
};
const shouldLogUpdateTiming = (durationMs) => {
    return process.env.DEV_TIMING === 'true' || durationMs >= SLOW_UPDATE_LOG_THRESHOLD_MS;
};
const addUpdateTimingLog = (filename, category, durationMs) => {
    if (!shouldLogUpdateTiming(durationMs)) {
        return;
    }
    const categoryLabel = category == null ? 'uncategorized' : category;
    addChangeLog(_jsx(InfoLog, { message: `processed ${filename} in ${formatElapsedSeconds(durationMs)} (${categoryLabel})` }));
};
const listener = (triggerRefresh, options = {}) => {
    const mintIgnoreGlobs = getMintIgnoreGlobs();
    chokidar
        .watch(CMD_EXEC_PATH, {
        ignoreInitial: true,
        ignored: (filePath) => {
            const relativePath = pathUtil.isAbsolute(filePath)
                ? pathUtil.relative(CMD_EXEC_PATH, filePath)
                : filePath;
            if (!relativePath) {
                return false;
            }
            return isMintIgnored(relativePath, mintIgnoreGlobs);
        },
        cwd: CMD_EXEC_PATH,
    })
        .on('add', (filename) => onAddEvent(filename, triggerRefresh, options))
        .on('change', (filename) => onChangeEvent(filename, triggerRefresh, options))
        .on('unlink', (filename) => onUnlinkEvent(filename, triggerRefresh, options));
};
const onAddEvent = async (filename, triggerRefresh, options) => {
    if (isMintIgnored(filename, getMintIgnoreGlobs())) {
        return;
    }
    try {
        const startTime = process.hrtime.bigint();
        const category = await onUpdateEvent(filename, triggerRefresh, options);
        addUpdateTimingLog(filename, category, getElapsedMs(startTime));
        if (category !== undefined) {
            addChangeLog(_jsx(AddedLog, { filename: filename }));
        }
    }
    catch (error) {
        console.error(error.message);
    }
};
const onChangeEvent = async (filename, triggerRefresh, options) => {
    if (isMintIgnored(filename, getMintIgnoreGlobs())) {
        return;
    }
    try {
        const startTime = process.hrtime.bigint();
        const category = await onUpdateEvent(filename, triggerRefresh, options);
        addUpdateTimingLog(filename, category, getElapsedMs(startTime));
        if (category !== undefined) {
            addChangeLog(_jsx(EditedLog, { filename: filename }));
        }
    }
    catch (error) {
        console.error(error.message);
    }
};
const onUnlinkEvent = async (filename, triggerRefresh, options) => {
    if (isMintIgnored(filename, getMintIgnoreGlobs())) {
        return;
    }
    try {
        const importedFiles = getImportedFilesFromCache();
        const potentialCategory = getFileCategory(filename, { importedFiles });
        frontmatterHashes.delete(filename);
        removePageMetadataCacheEntryForFile(filename);
        if (hasTrackedReferencedFile(filename)) {
            try {
                const { mintConfig, openApiFiles, docsConfig } = await getDocsState(handleParseError);
                await refreshTrackedReferencedFiles();
                if (mintConfig) {
                    await MintConfigUpdater.writeConfigFile(mintConfig, CLIENT_PATH);
                }
                await DocsConfigUpdater.writeConfigFile(docsConfig, CLIENT_PATH);
                await updateOpenApiFiles(openApiFiles, suppressParseError);
                await updateCustomLanguages(docsConfig, suppressParseError);
                await updateGeneratedNav(suppressParseError);
            }
            catch (err) {
                if (getCurrentErrorLogs().length === 0) {
                    console.error(err);
                }
            }
            const updatedSnippets = await regenerateAllSnippets();
            await generatePagesWithImports(new Set(updatedSnippets));
            const importChanges = removeFromImportCache(filename);
            await syncImportedFileLocations(importChanges);
            triggerRefresh();
            addChangeLog(_jsx(DeletedLog, { filename: filename }));
            return;
        }
        if (potentialCategory == null) {
            const importChanges = removeFromImportCache(filename);
            await syncImportedFileLocations(importChanges);
            return;
        }
        const targetPath = getTargetPath(potentialCategory, filename);
        if (potentialCategory === 'page' ||
            potentialCategory === 'snippet' ||
            potentialCategory === 'mintConfig' ||
            potentialCategory === 'docsConfig' ||
            potentialCategory === 'staticFile' ||
            potentialCategory === 'snippet-v2' ||
            potentialCategory === 'css' ||
            potentialCategory === 'js' ||
            potentialCategory === 'generatedStaticFile' ||
            potentialCategory === 'skillFile') {
            await fse.remove(targetPath);
        }
        switch (potentialCategory) {
            case 'mintConfig':
                console.error('mint.json has been deleted.');
                await validateConfigFiles();
                break;
            case 'docsConfig':
                console.error('docs.json has been deleted.');
                await validateConfigFiles();
                break;
            case 'mintIgnore':
                addChangeLog(_jsx(WarningLog, { message: ".mintignore has been deleted. Rebuilding..." }));
                try {
                    await fse.emptyDir(NEXT_PUBLIC_PATH);
                    await fse.emptyDir(NEXT_PROPS_PATH);
                    const prebuildResult = await prebuild(CMD_EXEC_PATH, options);
                    await initializeImportCache(CMD_EXEC_PATH, prebuildResult?.fileImportsMap);
                    await initializeFrontmatterHashCache();
                    await refreshTrackedReferencedFiles();
                }
                catch (err) {
                    console.error('Error rebuilding after .mintignore deletion:', err);
                }
                break;
        }
        const importChanges = removeFromImportCache(filename);
        await syncImportedFileLocations(importChanges);
        addChangeLog(_jsx(DeletedLog, { filename: filename }));
    }
    catch (error) {
        console.error(error.message);
    }
};
const getTargetPath = (potentialCategory, filePath) => {
    switch (potentialCategory) {
        case 'page':
            return pathUtil.join(NEXT_PROPS_PATH, filePath);
        case 'mintConfig':
            return pathUtil.join(NEXT_PROPS_PATH, 'mint.json');
        case 'docsConfig':
            return pathUtil.join(NEXT_PROPS_PATH, 'docs.json');
        case 'potentialYamlOpenApiSpec':
        case 'potentialJsonOpenApiSpec':
            return pathUtil.join(NEXT_PROPS_PATH, 'openApiFiles.json');
        case 'generatedStaticFile':
            return pathUtil.join(NEXT_PUBLIC_PATH, filePath);
        case 'mintIgnore':
            return pathUtil.join(NEXT_PROPS_PATH, 'mint-ignore.json');
        case 'snippet':
        case 'staticFile':
        case 'snippet-v2':
        case 'css':
        case 'js':
            return pathUtil.join(NEXT_PUBLIC_PATH, filePath);
        case 'skillFile':
            return pathUtil.join(NEXT_PUBLIC_PATH, filePath);
        default:
            throw new Error('Invalid category');
    }
};
const validateConfigFiles = async () => {
    try {
        const mintConfigPath = pathUtil.join(CMD_EXEC_PATH, 'mint.json');
        const docsConfigPath = pathUtil.join(CMD_EXEC_PATH, 'docs.json');
        const mintConfigExists = await fse.pathExists(mintConfigPath);
        const docsConfigExists = await fse.pathExists(docsConfigPath);
        if (!mintConfigExists && !docsConfigExists) {
            console.error('⚠️ Error: Neither mint.json nor docs.json found in the directory');
            process.exit(1);
        }
    }
    catch (error) {
        console.error('⚠️ Error validating configuration files:', error);
    }
};
const initializeFrontmatterHashCache = async () => {
    frontmatterHashes.clear();
    clearPageMetadataCache();
    const importedFiles = getImportedFilesFromCache();
    const pageFilenames = getFileListSync(CMD_EXEC_PATH).filter((file) => {
        return getFileCategory(file, { importedFiles }) === 'page';
    });
    await Promise.all(pageFilenames.map(async (filename) => {
        try {
            const filePath = pathUtil.join(CMD_EXEC_PATH, filename);
            const contentStr = (await readFile(filePath)).toString();
            seedFrontmatterHashForPage(filename, contentStr, frontmatterHashes, suppressParseError);
            upsertPageMetadataCacheEntryForFile(filename, contentStr);
        }
        catch { }
    }));
};
const initializeDocsConfigHashCache = async () => {
    clearDocsConfigCaches();
    // Seed the cache from the raw config (no OpenAPI generation) so it stays
    // consistent with the comparison done on each docs.json save.
    const docsJsonPath = pathUtil.join(CMD_EXEC_PATH, 'docs.json');
    try {
        if (await fse.pathExists(docsJsonPath)) {
            const docsConfig = await DocsConfigUpdater.getConfig(docsJsonPath, false, CMD_EXEC_PATH, suppressParseError);
            updateDocsConfigHashCache(docsConfig);
            return;
        }
    }
    catch { }
};
/**
 * This function is called when a file is added or changed
 * @param filename
 * @returns FileCategory
 */
const onUpdateEvent = async (filename, triggerRefresh, options = {}) => {
    clearErrorLogs();
    const filePath = pathUtil.join(CMD_EXEC_PATH, filename);
    const importChanges = await updateImportCacheForFile(CMD_EXEC_PATH, filename);
    await syncImportedFileLocations(importChanges);
    const importedFiles = getImportedFilesFromCache();
    const potentialCategory = getFileCategory(filename, { importedFiles });
    if (hasTrackedReferencedFile(filename)) {
        try {
            const { mintConfig, openApiFiles, docsConfig, pagesAcc } = await getDocsState(handleParseError);
            await refreshTrackedReferencedFiles();
            if (mintConfig) {
                await MintConfigUpdater.writeConfigFile(mintConfig, CLIENT_PATH);
            }
            await DocsConfigUpdater.writeConfigFile(docsConfig, CLIENT_PATH);
            await updateOpenApiFiles(openApiFiles, suppressParseError);
            await updateCustomLanguages(docsConfig, suppressParseError);
            await updateGeneratedNav(suppressParseError, { pagesAcc, docsConfig });
        }
        catch (err) {
            if (getCurrentErrorLogs().length === 0) {
                console.error(err);
            }
        }
        const updatedSnippets = await regenerateAllSnippets();
        await generatePagesWithImports(new Set(updatedSnippets));
        // A tracked $ref change rebuilds config-derived artifacts; keep every docs
        // config hash in sync with the same raw docs.json source used by save checks.
        const rawDocsConfig = await getCurrentRawDocsConfig();
        if (rawDocsConfig) {
            updateDocsConfigHashCache(rawDocsConfig);
        }
        triggerRefresh();
        return null;
    }
    if (potentialCategory == null) {
        triggerRefresh();
        return undefined;
    }
    let regenerateNav = false;
    let regenerateNavFromPageMetadataCache = false;
    let category = potentialCategory === 'potentialYamlOpenApiSpec' ||
        potentialCategory === 'potentialJsonOpenApiSpec'
        ? 'staticFile'
        : potentialCategory;
    const targetPath = getTargetPath(potentialCategory, filename);
    switch (potentialCategory) {
        case 'page': {
            let contentStr = (await readFile(filePath)).toString();
            regenerateNavFromPageMetadataCache = await shouldRegenerateNavForPage(filename, contentStr, frontmatterHashes, handleParseError);
            const tree = await preparseMdxTree(contentStr, CMD_EXEC_PATH, filePath, suppressParseError);
            const importsResponse = await findAndRemoveImports(tree);
            if (hasImports(importsResponse)) {
                contentStr = stringifyTree(await resolvePageImports({ ...importsResponse, filename }));
            }
            const { pageContent, pageMetadata, slug } = await createPage(filename, contentStr, CMD_EXEC_PATH, [], [], handleParseError);
            await fse.outputFile(targetPath, pageContent, {
                flag: 'w',
            });
            upsertPageMetadataCacheEntry(slug, pageMetadata);
            break;
        }
        case 'snippet': {
            await fse.copy(filePath, targetPath);
            break;
        }
        case 'snippet-v2': {
            const variables = await getCurrentVariables();
            let contentStr = replaceVariables((await readFile(filePath)).toString(), variables);
            const tree = await preparseMdxTree(contentStr, CMD_EXEC_PATH, filePath, handleParseError);
            const importsResponse = await findAndRemoveImports(tree);
            if (hasImports(importsResponse)) {
                contentStr = stringifyTree(await resolveImportsFromImportClosure({ ...importsResponse, filename }));
            }
            await fse.outputFile(targetPath, contentStr, {
                flag: 'w',
            });
            const updatedSnippets = await generateDependentSnippets(filename, importsResponse);
            await generatePagesWithImports(new Set(updatedSnippets));
            break;
        }
        case 'mintConfig':
        case 'docsConfig': {
            const { valid, error } = isJsonValid(filePath);
            if (!valid) {
                addErrorLog(_jsx(ErrorLog, { message: `Syntax error in ${filename}: ${error}` }));
                return undefined;
            }
            if (potentialCategory === 'docsConfig') {
                try {
                    // Parse the raw config only (cheap) so we can decide what actually
                    // needs rebuilding before paying for the expensive OpenAPI/nav work.
                    const docsConfig = await DocsConfigUpdater.getConfig(filePath, false, CMD_EXEC_PATH, handleParseError);
                    const newApiHash = getDocsConfigApiHash(docsConfig);
                    const apiChanged = hasDocsConfigApiChanged(newApiHash);
                    const newNavigationHash = getDocsConfigNavigationHash(docsConfig);
                    const navigationChanged = hasDocsConfigNavigationChanged(newNavigationHash);
                    const newVariablesHash = getVariablesHash(docsConfig.variables);
                    const variablesChanged = hasDocsConfigVariablesChanged(newVariablesHash);
                    const shouldUseNavigationFastPath = navigationChanged && !apiChanged && !hasDocsConfigApiRoutes(docsConfig);
                    const shouldRegenerateGeneratedData = apiChanged || (navigationChanged && hasDocsConfigApiRoutes(docsConfig));
                    await refreshTrackedReferencedFiles();
                    if (shouldUseNavigationFastPath) {
                        // Plain MDX navigation changed without API changes: regenerate
                        // generatedDocsNav from cached page metadata. If a page references a
                        // spec via frontmatter, this falls back to the full rebuild and
                        // returns the API-expanded config, which we then persist.
                        await updateCustomLanguages(docsConfig, suppressParseError);
                        const generatedDocsConfig = await updateGeneratedNavFromPageMetadataCache(docsConfig, suppressParseError);
                        await DocsConfigUpdater.writeConfigFile(generatedDocsConfig, CLIENT_PATH);
                    }
                    else if (shouldRegenerateGeneratedData) {
                        // Preserve the old full-rebuild route for top-level api, API-derived
                        // navigation, and top-level api removals.
                        const docsState = await getDocsState(handleParseError);
                        await DocsConfigUpdater.writeConfigFile(docsState.docsConfig, CLIENT_PATH);
                        await updateOpenApiFiles(docsState.openApiFiles, suppressParseError);
                        await updateCustomLanguages(docsState.docsConfig, suppressParseError);
                        await updateGeneratedNav(suppressParseError, {
                            pagesAcc: docsState.pagesAcc,
                            docsConfig: docsState.docsConfig,
                        });
                    }
                    else {
                        // Non-generated docs.json changes (theme, styling, metadata, etc.) do
                        // not need OpenAPI/nav regeneration. Preserve the current generated
                        // navigation so API-expanded docs config does not collapse to raw nav.
                        await updateCustomLanguages(docsConfig, suppressParseError);
                        await writeDocsConfigWithCurrentNavigation(docsConfig);
                    }
                    if (variablesChanged) {
                        const updatedSnippets = await regenerateAllSnippets();
                        await generatePagesWithImports(new Set(updatedSnippets));
                    }
                    // Update the cache only after every step above succeeded, so a failure
                    // mid-update doesn't leave the cache claiming work that never happened.
                    updateDocsConfigHashCache(docsConfig);
                }
                catch (err) {
                    if (getCurrentErrorLogs().length > 0) {
                        // no-op to suppress duplicate error logging
                        return undefined;
                    }
                    else {
                        console.error(err);
                    }
                }
                break;
            }
            // mint.json: preserve the original full-rebuild behavior.
            regenerateNav = true;
            try {
                const { mintConfig, openApiFiles, docsConfig } = await getDocsState(handleParseError);
                await refreshTrackedReferencedFiles();
                if (mintConfig) {
                    await MintConfigUpdater.writeConfigFile(mintConfig, CLIENT_PATH);
                }
                await DocsConfigUpdater.writeConfigFile(docsConfig, CLIENT_PATH);
                await updateOpenApiFiles(openApiFiles, suppressParseError);
                await updateCustomLanguages(docsConfig, suppressParseError);
            }
            catch (err) {
                if (getCurrentErrorLogs().length > 0) {
                    // no-op to suppress duplicate error logging
                    return undefined;
                }
                else {
                    console.error(err);
                }
            }
            break;
        }
        case 'mintIgnore': {
            try {
                addChangeLog(_jsx(InfoLog, { message: ".mintignore has been updated. Rebuilding..." }));
                await fse.emptyDir(NEXT_PUBLIC_PATH);
                await fse.emptyDir(NEXT_PROPS_PATH);
                const prebuildResult = await prebuild(CMD_EXEC_PATH, options);
                await initializeImportCache(CMD_EXEC_PATH, prebuildResult?.fileImportsMap);
                await initializeFrontmatterHashCache();
                await refreshTrackedReferencedFiles();
            }
            catch (err) {
                console.error(err.message);
            }
            break;
        }
        case 'potentialYamlOpenApiSpec':
        case 'potentialJsonOpenApiSpec': {
            let doc;
            try {
                const file = await fs.readFile(filePath, 'utf-8');
                const { schema } = await validate(yaml.load(file));
                doc = schema;
            }
            catch {
                doc = undefined;
            }
            if (doc) {
                await upsertOpenApiFile({
                    filename: pathUtil.parse(filename).name,
                    originalFileLocation: '/' + filename,
                    spec: doc,
                });
                await updateOpenApiFiles();
                regenerateNav = true;
                category = 'openApi';
            }
            else {
                const customLanguages = JSON.parse(await fs.readFile(pathUtil.join(NEXT_PROPS_PATH, 'customLanguages.json'), 'utf-8'));
                let updated = false;
                for (const [index, customLanguage] of customLanguages.entries()) {
                    if (filePath.endsWith(customLanguage.filePath)) {
                        updated = true;
                        const file = await fs.readFile(filePath, 'utf-8');
                        customLanguages[index] = { content: file, filePath: customLanguage.filePath };
                        break;
                    }
                }
                if (updated) {
                    await fs.writeFile(pathUtil.join(NEXT_PROPS_PATH, 'customLanguages.json'), JSON.stringify(customLanguages, null, 2));
                }
            }
            break;
        }
        case 'skillFile':
        case 'css':
        case 'js':
        case 'generatedStaticFile':
        case 'staticFile': {
            if (await isFileSizeValid(filePath, 20)) {
                await fse.copy(filePath, targetPath);
            }
            else {
                console.error(Chalk.red(`🚨 The file at ${filename} is too big. The maximum file size is 20 mb.`));
            }
            break;
        }
    }
    if (regenerateNavFromPageMetadataCache) {
        const rawDocsConfig = await getCurrentRawDocsConfig();
        const docsConfig = rawDocsConfig ?? (await getCurrentDocsConfig());
        if (hasDocsConfigApiRoutes(docsConfig)) {
            const generatedDocsConfig = await updateGeneratedNav(suppressParseError);
            await DocsConfigUpdater.writeConfigFile(generatedDocsConfig, CLIENT_PATH);
        }
        else {
            await updateGeneratedNavFromPageMetadataCache(docsConfig, suppressParseError);
        }
    }
    else if (regenerateNav) {
        // TODO: Instead of re-generating the entire nav, optimize by just updating the specific page that changed.
        const docsConfig = await updateGeneratedNav(suppressParseError);
        await DocsConfigUpdater.writeConfigFile(docsConfig, CLIENT_PATH);
    }
    triggerRefresh();
    return category;
};
export { initializeDocsConfigHashCache, initializeFrontmatterHashCache, initializeImportCache };
export default listener;
