// @mintlify/previewing/dist/local-preview/listener/page-metadata-cache.js
import { getAllPathsInDocsNav, getDecoratedNavPageAndSlug } from '@mintlify/common';
import { generateDecoratedDocsNavigationFromPages } from '@mintlify/prebuild';
import fse from 'fs-extra';
import pathUtil from 'path';
import { CMD_EXEC_PATH } from '../../constants.js';
const pageMetadataBySlug = {};
const getPossiblePageFilenames = (pagePath) => {
    const normalizedPath = pagePath.startsWith('/') ? pagePath.slice(1) : pagePath;
    if (normalizedPath.endsWith('.mdx') || normalizedPath.endsWith('.md')) {
        return [normalizedPath];
    }
    return [`${normalizedPath}.mdx`, `${normalizedPath}.md`];
};
export const clearPageMetadataCache = () => {
    for (const slug of Object.keys(pageMetadataBySlug)) {
        delete pageMetadataBySlug[slug];
    }
};
export const upsertPageMetadataCacheEntry = (slug, pageMetadata) => {
    pageMetadataBySlug[slug] = pageMetadata;
};
export const removePageMetadataCacheEntry = (slug) => {
    delete pageMetadataBySlug[slug];
};
export const removePageMetadataCacheEntryForFile = (filename) => {
    removePageMetadataCacheEntry(filename.replace(/\.mdx?$/, ''));
};
// Returns true when the page references an OpenAPI/AsyncAPI spec via frontmatter.
// Those titles are derived from the schema, which the cache (built without spec
// files) cannot reproduce, so callers route such pages through the full rebuild.
export const upsertPageMetadataCacheEntryForFile = (filename, contentStr) => {
    const { slug, pageMetadata } = getDecoratedNavPageAndSlug(filename, contentStr, [], []);
    upsertPageMetadataCacheEntry(slug, pageMetadata);
    return pageMetadata.openapi != null || pageMetadata.asyncapi != null;
};
// Returns true when any page in the navigation references an OpenAPI/AsyncAPI
// spec via frontmatter.
export const ensurePageMetadataCacheForDocsConfig = async (docsConfig) => {
    const navPaths = getAllPathsInDocsNav(docsConfig.navigation);
    const results = await Promise.all(navPaths.map(async (navPath) => {
        for (const filename of getPossiblePageFilenames(navPath)) {
            const filePath = pathUtil.join(CMD_EXEC_PATH, filename);
            if (!(await fse.pathExists(filePath))) {
                continue;
            }
            const contentStr = (await fse.readFile(filePath)).toString();
            return upsertPageMetadataCacheEntryForFile(filename, contentStr);
        }
        return false;
    }));
    return results.some((hasApiSpecPage) => hasApiSpecPage);
};
export const generateDocsNavFromPageMetadataCache = (docsConfig) => {
    return generateDecoratedDocsNavigationFromPages(pageMetadataBySlug, docsConfig.navigation);
};
