// @mintlify/previewing/dist/local-preview/listener/resolve-imports-from-import-closure.js
import { findAndRemoveImports, hasImports, optionallyAddLeadingSlash, optionallyRemoveLeadingSlash, replaceVariables, resolveAllImports as baseResolveAllImports, resolveImportPath, topologicalSort, } from '@mintlify/common';
import { preparseMdxTree } from '@mintlify/prebuild';
import fse from 'fs-extra';
import { join } from 'path';
import { CMD_EXEC_PATH, NEXT_PUBLIC_PATH } from '../../constants.js';
import { getCurrentVariables, handleParseError } from './utils.js';
const getImportSourcePath = async (filename) => {
    const generatedPath = join(NEXT_PUBLIC_PATH, filename);
    if (await fse.pathExists(generatedPath)) {
        return {
            baseDir: NEXT_PUBLIC_PATH,
            path: generatedPath,
        };
    }
    const originalPath = join(CMD_EXEC_PATH, filename);
    if (await fse.pathExists(originalPath)) {
        return {
            baseDir: CMD_EXEC_PATH,
            path: originalPath,
        };
    }
    return null;
};
const parseSnippetWithImports = async ({ filename, variables, snippetsByFilename, }) => {
    const normalizedFilename = optionallyAddLeadingSlash(filename);
    if (snippetsByFilename.has(normalizedFilename)) {
        return;
    }
    const importSource = await getImportSourcePath(optionallyRemoveLeadingSlash(normalizedFilename));
    if (importSource == null) {
        return;
    }
    const content = replaceVariables((await fse.readFile(importSource.path)).toString(), variables);
    const tree = await preparseMdxTree(content, importSource.baseDir, importSource.path, handleParseError);
    const importData = await findAndRemoveImports(tree);
    const fileWithImports = {
        filename: normalizedFilename,
        ...importData,
    };
    snippetsByFilename.set(normalizedFilename, fileWithImports);
    await Promise.all(Object.keys(importData.importMap).map(async (source) => {
        const resolvedPath = resolveImportPath(source, normalizedFilename);
        if (resolvedPath == null) {
            return;
        }
        await parseSnippetWithImports({
            filename: resolvedPath,
            variables,
            snippetsByFilename,
        });
    }));
};
const resolveSnippetImports = async (snippetsByFilename) => {
    const graph = {};
    for (const snippet of snippetsByFilename.values()) {
        graph[snippet.filename] = Object.keys(snippet.importMap)
            .map((source) => resolveImportPath(source, snippet.filename))
            .filter((resolvedPath) => {
            return resolvedPath != null && snippetsByFilename.has(resolvedPath);
        });
    }
    const orderedFilenames = topologicalSort(graph).reverse();
    const resolvedSnippets = [];
    for (const filename of orderedFilenames) {
        const snippet = snippetsByFilename.get(filename);
        if (snippet == null) {
            continue;
        }
        const tree = hasImports(snippet)
            ? await baseResolveAllImports({
                snippets: resolvedSnippets,
                fileWithImports: snippet,
            })
            : snippet.tree;
        resolvedSnippets.push({ filename: snippet.filename, tree });
    }
    return resolvedSnippets;
};
export const resolveImportsFromImportClosure = async (fileWithImports) => {
    const variables = await getCurrentVariables();
    const snippetsByFilename = new Map();
    await Promise.all(Object.keys(fileWithImports.importMap).map(async (source) => {
        const resolvedPath = resolveImportPath(source, fileWithImports.filename);
        if (resolvedPath == null) {
            return;
        }
        await parseSnippetWithImports({
            filename: resolvedPath,
            variables,
            snippetsByFilename,
        });
    }));
    const snippets = await resolveSnippetImports(snippetsByFilename);
    return await baseResolveAllImports({
        snippets,
        fileWithImports,
    });
};
