// @mintlify/models/dist/types/stytchRbacPolicy.js
// This file is autogenerated with 'yarn update-stytch-types'. It will overwrite any changes you've made here.
export const ACTIONS = ['create', 'delete', 'read', 'update', 'upsert', 'write'];
export const ORG_RESOURCES = [
    'org',
    'org.apiKey',
    'org.auditLogs',
    'org.billing',
    'org.deployments',
    'org.editorPermissions',
    'org.gitlab',
    'org.growthData',
    'org.members',
    'org.scim',
    'org.securityContact',
    'org.sso',
];
export const USER_RESOURCES = [
    'user',
    'discord.integration',
    'github.integration',
    'gitlab.integration',
    'user.role',
];
export const DEPLOYMENT_RESOURCES = [
    'agent',
    'autopilot',
    'config',
    'deployment',
    'docs',
    'nav',
    'session',
    'deployment.agents',
    'deployment.analytics',
    'deployment.assistant',
    'deployment.auth',
    'deployment.basePath',
    'deployment.customDomain',
    'deployment.discoveryApiKey',
    'deployment.endUserAuth',
    'deployment.gitSource',
    'deployment.history',
    'deployment.pdf',
    'deployment.personalization',
    'deployment.preview',
    'deployment.quota',
    'deployment.relatedPages',
    'deployment.slack',
    'deployment.sourceChecks',
    'deployment.translations',
    'deployment.trial',
    'deployment.usage',
    'git.content',
    'git.write',
    'mcp.analytics',
    'mcp.billing',
    'mcp.deployment',
    'mcp.integrations',
    'mcp.members',
    'slack.integration',
];
export const ALL_PERMISSIONS = [
    'agent.create',
    'agent.read',
    'agent.update',
    'autopilot.create',
    'autopilot.delete',
    'autopilot.read',
    'autopilot.update',
    'config.write',
    'deployment.agents.read',
    'deployment.analytics.delete',
    'deployment.analytics.read',
    'deployment.analytics.update',
    'deployment.assistant.create',
    'deployment.assistant.delete',
    'deployment.assistant.read',
    'deployment.assistant.update',
    'deployment.auth.create',
    'deployment.auth.delete',
    'deployment.auth.read',
    'deployment.auth.update',
    'deployment.auth.upsert',
    'deployment.basePath.create',
    'deployment.basePath.delete',
    'deployment.basePath.read',
    'deployment.basePath.update',
    'deployment.basePath.upsert',
    'deployment.create',
    'deployment.customDomain.create',
    'deployment.customDomain.delete',
    'deployment.customDomain.read',
    'deployment.customDomain.update',
    'deployment.customDomain.upsert',
    'deployment.delete',
    'deployment.discoveryApiKey.create',
    'deployment.discoveryApiKey.delete',
    'deployment.discoveryApiKey.read',
    'deployment.endUserAuth.read',
    'deployment.gitSource.create',
    'deployment.gitSource.delete',
    'deployment.gitSource.read',
    'deployment.gitSource.update',
    'deployment.gitSource.upsert',
    'deployment.history.delete',
    'deployment.history.read',
    'deployment.pdf.create',
    'deployment.pdf.delete',
    'deployment.pdf.read',
    'deployment.personalization.create',
    'deployment.personalization.delete',
    'deployment.personalization.read',
    'deployment.personalization.update',
    'deployment.personalization.upsert',
    'deployment.preview.create',
    'deployment.preview.read',
    'deployment.preview.update',
    'deployment.quota.read',
    'deployment.quota.update',
    'deployment.read',
    'deployment.relatedPages.read',
    'deployment.relatedPages.update',
    'deployment.slack.read',
    'deployment.sourceChecks.read',
    'deployment.sourceChecks.update',
    'deployment.translations.create',
    'deployment.translations.read',
    'deployment.translations.update',
    'deployment.trial.read',
    'deployment.update',
    'deployment.upsert',
    'deployment.usage.read',
    'discord.integration.create',
    'discord.integration.delete',
    'discord.integration.read',
    'discord.integration.update',
    'docs.read',
    'docs.write',
    'git.content.read',
    'git.write.create',
    'git.write.delete',
    'git.write.read',
    'git.write.update',
    'github.integration.create',
    'github.integration.delete',
    'github.integration.read',
    'gitlab.integration.read',
    'mcp.analytics.read',
    'mcp.billing.read',
    'mcp.deployment.read',
    'mcp.deployment.write',
    'mcp.integrations.write',
    'mcp.members.read',
    'mcp.members.write',
    'nav.write',
    'org.apiKey.create',
    'org.apiKey.delete',
    'org.apiKey.read',
    'org.auditLogs.read',
    'org.billing.read',
    'org.billing.update',
    'org.delete',
    'org.deployments.read',
    'org.editorPermissions.read',
    'org.editorPermissions.update',
    'org.gitlab.create',
    'org.gitlab.read',
    'org.gitlab.update',
    'org.growthData.update',
    'org.members.create',
    'org.members.read',
    'org.read',
    'org.scim.create',
    'org.scim.delete',
    'org.scim.read',
    'org.scim.update',
    'org.securityContact.delete',
    'org.securityContact.read',
    'org.securityContact.upsert',
    'org.sso.create',
    'org.sso.delete',
    'org.sso.read',
    'org.sso.update',
    'org.update',
    'session.write',
    'slack.integration.create',
    'slack.integration.delete',
    'slack.integration.read',
    'slack.integration.update',
    'user.delete',
    'user.read',
    'user.role.create',
    'user.role.delete',
    'user.role.read',
    'user.role.update',
    'user.update',
];
// Permissions are cascaded: admin has editor+viewer, editor has viewer
export const ROLE_PERMISSIONS = {
    admin: new Set([
        'agent.create',
        'agent.read',
        'agent.update',
        'autopilot.create',
        'autopilot.delete',
        'autopilot.read',
        'autopilot.update',
        'config.write',
        'deployment.agents.read',
        'deployment.analytics.delete',
        'deployment.analytics.read',
        'deployment.analytics.update',
        'deployment.assistant.create',
        'deployment.assistant.delete',
        'deployment.assistant.read',
        'deployment.assistant.update',
        'deployment.auth.create',
        'deployment.auth.delete',
        'deployment.auth.read',
        'deployment.auth.update',
        'deployment.auth.upsert',
        'deployment.basePath.create',
        'deployment.basePath.delete',
        'deployment.basePath.read',
        'deployment.basePath.update',
        'deployment.basePath.upsert',
        'deployment.create',
        'deployment.customDomain.create',
        'deployment.customDomain.delete',
        'deployment.customDomain.read',
        'deployment.customDomain.update',
        'deployment.customDomain.upsert',
        'deployment.delete',
        'deployment.discoveryApiKey.create',
        'deployment.discoveryApiKey.delete',
        'deployment.discoveryApiKey.read',
        'deployment.endUserAuth.read',
        'deployment.gitSource.create',
        'deployment.gitSource.delete',
        'deployment.gitSource.read',
        'deployment.gitSource.update',
        'deployment.gitSource.upsert',
        'deployment.history.delete',
        'deployment.history.read',
        'deployment.pdf.create',
        'deployment.pdf.delete',
        'deployment.pdf.read',
        'deployment.personalization.create',
        'deployment.personalization.delete',
        'deployment.personalization.read',
        'deployment.personalization.update',
        'deployment.personalization.upsert',
        'deployment.preview.create',
        'deployment.preview.read',
        'deployment.preview.update',
        'deployment.quota.read',
        'deployment.quota.update',
        'deployment.read',
        'deployment.relatedPages.read',
        'deployment.relatedPages.update',
        'deployment.slack.read',
        'deployment.sourceChecks.read',
        'deployment.sourceChecks.update',
        'deployment.translations.create',
        'deployment.translations.read',
        'deployment.translations.update',
        'deployment.trial.read',
        'deployment.update',
        'deployment.upsert',
        'deployment.usage.read',
        'discord.integration.create',
        'discord.integration.delete',
        'discord.integration.read',
        'discord.integration.update',
        'docs.read',
        'docs.write',
        'git.content.read',
        'git.write.create',
        'git.write.delete',
        'git.write.read',
        'git.write.update',
        'github.integration.create',
        'github.integration.delete',
        'github.integration.read',
        'gitlab.integration.read',
        'mcp.analytics.read',
        'mcp.billing.read',
        'mcp.deployment.read',
        'mcp.deployment.write',
        'mcp.integrations.write',
        'mcp.members.read',
        'mcp.members.write',
        'nav.write',
        'org.apiKey.create',
        'org.apiKey.delete',
        'org.apiKey.read',
        'org.auditLogs.read',
        'org.billing.read',
        'org.billing.update',
        'org.delete',
        'org.deployments.read',
        'org.editorPermissions.read',
        'org.editorPermissions.update',
        'org.gitlab.create',
        'org.gitlab.read',
        'org.gitlab.update',
        'org.growthData.update',
        'org.members.create',
        'org.members.read',
        'org.read',
        'org.scim.create',
        'org.scim.delete',
        'org.scim.read',
        'org.scim.update',
        'org.securityContact.delete',
        'org.securityContact.read',
        'org.securityContact.upsert',
        'org.sso.create',
        'org.sso.delete',
        'org.sso.read',
        'org.sso.update',
        'org.update',
        'session.write',
        'slack.integration.create',
        'slack.integration.delete',
        'slack.integration.read',
        'slack.integration.update',
        'user.delete',
        'user.read',
        'user.role.create',
        'user.role.delete',
        'user.role.read',
        'user.role.update',
        'user.update',
    ]),
    editor: new Set([
        'agent.create',
        'agent.read',
        'agent.update',
        'autopilot.create',
        'autopilot.delete',
        'autopilot.read',
        'autopilot.update',
        'config.write',
        'deployment.agents.read',
        'deployment.analytics.delete',
        'deployment.analytics.read',
        'deployment.analytics.update',
        'deployment.assistant.create',
        'deployment.assistant.delete',
        'deployment.assistant.read',
        'deployment.assistant.update',
        'deployment.auth.create',
        'deployment.auth.delete',
        'deployment.auth.read',
        'deployment.auth.update',
        'deployment.create',
        'deployment.customDomain.read',
        'deployment.discoveryApiKey.create',
        'deployment.discoveryApiKey.delete',
        'deployment.discoveryApiKey.read',
        'deployment.endUserAuth.read',
        'deployment.history.delete',
        'deployment.history.read',
        'deployment.pdf.create',
        'deployment.pdf.delete',
        'deployment.pdf.read',
        'deployment.preview.create',
        'deployment.preview.read',
        'deployment.preview.update',
        'deployment.quota.read',
        'deployment.read',
        'deployment.relatedPages.read',
        'deployment.relatedPages.update',
        'deployment.slack.read',
        'deployment.sourceChecks.read',
        'deployment.sourceChecks.update',
        'deployment.translations.create',
        'deployment.translations.read',
        'deployment.translations.update',
        'deployment.trial.read',
        'deployment.update',
        'deployment.usage.read',
        'discord.integration.create',
        'discord.integration.delete',
        'discord.integration.read',
        'discord.integration.update',
        'docs.read',
        'docs.write',
        'git.content.read',
        'git.write.create',
        'git.write.delete',
        'git.write.read',
        'git.write.update',
        'github.integration.create',
        'github.integration.delete',
        'github.integration.read',
        'gitlab.integration.read',
        'mcp.analytics.read',
        'mcp.deployment.read',
        'mcp.deployment.write',
        'mcp.integrations.write',
        'mcp.members.read',
        'nav.write',
        'org.apiKey.create',
        'org.apiKey.delete',
        'org.apiKey.read',
        'org.billing.read',
        'org.deployments.read',
        'org.editorPermissions.read',
        'org.gitlab.create',
        'org.gitlab.read',
        'org.gitlab.update',
        'org.growthData.update',
        'org.members.create',
        'org.members.read',
        'org.read',
        'org.scim.read',
        'org.update',
        'session.write',
        'slack.integration.create',
        'slack.integration.delete',
        'slack.integration.read',
        'slack.integration.update',
        'user.read',
        'user.role.create',
        'user.role.read',
        'user.update',
    ]),
    viewer: new Set([
        'agent.read',
        'autopilot.read',
        'deployment.agents.read',
        'deployment.analytics.read',
        'deployment.assistant.read',
        'deployment.auth.read',
        'deployment.customDomain.read',
        'deployment.discoveryApiKey.read',
        'deployment.endUserAuth.read',
        'deployment.history.read',
        'deployment.pdf.create',
        'deployment.pdf.read',
        'deployment.preview.read',
        'deployment.quota.read',
        'deployment.read',
        'deployment.relatedPages.read',
        'deployment.slack.read',
        'deployment.sourceChecks.read',
        'deployment.translations.read',
        'deployment.trial.read',
        'deployment.usage.read',
        'discord.integration.read',
        'docs.read',
        'git.content.read',
        'git.write.read',
        'github.integration.read',
        'gitlab.integration.read',
        'mcp.analytics.read',
        'mcp.deployment.read',
        'mcp.members.read',
        'org.apiKey.read',
        'org.billing.read',
        'org.deployments.read',
        'org.editorPermissions.read',
        'org.gitlab.read',
        'org.growthData.update',
        'org.members.read',
        'org.read',
        'org.scim.read',
        'slack.integration.read',
        'user.read',
        'user.role.read',
        'user.update',
    ]),
};
