// @mintlify/validation/dist/mint-config/schemas/v2/properties/navigation/menu.js
import { z } from 'zod';
import { asyncApiSchema } from '../reusable/asyncapi.js';
import { hiddenSchema } from '../reusable/hidden.js';
import { hrefSchema } from '../reusable/href.js';
import { iconSchema } from '../reusable/icon.js';
import { openApiSchema } from '../reusable/openapi.js';
import { anchorsSchema, decoratedAnchorsSchema } from './anchors.js';
import { decoratedDropdownsSchema, dropdownsSchema } from './dropdown.js';
import { globalSchema } from './global.js';
import { decoratedGroupsSchema, directoryField, groupsSchema } from './groups.js';
import { decoratedLanguagesSchema, languagesSchema } from './languages.js';
import { decoratedPagesSchema, pagesSchema } from './pages.js';
import { decoratedProductsSchema, productsSchema } from './products.js';
import { decoratedTabsSchema, tabsSchema } from './tabs.js';
import { decoratedVersionsSchema, versionsSchema } from './version.js';
export const baseMenuItemSchema = z.object(Object.assign({ item: z.string().nonempty().describe('The name of the menu item'), icon: iconSchema.optional(), description: z.string().optional().describe('The description of the menu item'), hidden: hiddenSchema.optional() }, directoryField));
export const nonRecursiveMenuItemSchema = baseMenuItemSchema.extend({ href: hrefSchema });
export const menuItemSchema = z.union([
    baseMenuItemSchema.extend({
        href: hrefSchema,
        openapi: openApiSchema.optional(),
        asyncapi: asyncApiSchema.optional(),
        global: z.lazy(() => globalSchema).optional(),
    }),
    baseMenuItemSchema.extend({
        languages: z.lazy(() => languagesSchema),
        openapi: openApiSchema.optional(),
        asyncapi: asyncApiSchema.optional(),
        global: z.lazy(() => globalSchema).optional(),
    }),
    baseMenuItemSchema.extend({
        versions: z.lazy(() => versionsSchema),
        openapi: openApiSchema.optional(),
        asyncapi: asyncApiSchema.optional(),
        global: z.lazy(() => globalSchema).optional(),
    }),
    baseMenuItemSchema.extend({
        dropdowns: z.lazy(() => dropdownsSchema),
        openapi: openApiSchema.optional(),
        asyncapi: asyncApiSchema.optional(),
        global: z.lazy(() => globalSchema).optional(),
    }),
    baseMenuItemSchema.extend({
        anchors: z.lazy(() => anchorsSchema),
        openapi: openApiSchema.optional(),
        asyncapi: asyncApiSchema.optional(),
        global: z.lazy(() => globalSchema).optional(),
    }),
    baseMenuItemSchema.extend({
        products: z.lazy(() => productsSchema),
        openapi: openApiSchema.optional(),
        asyncapi: asyncApiSchema.optional(),
        global: z.lazy(() => globalSchema).optional(),
    }),
    baseMenuItemSchema.extend({
        tabs: z.lazy(() => tabsSchema),
        openapi: openApiSchema.optional(),
        asyncapi: asyncApiSchema.optional(),
        global: z.lazy(() => globalSchema).optional(),
    }),
    baseMenuItemSchema.extend({
        groups: z.lazy(() => groupsSchema),
        openapi: openApiSchema.optional(),
        asyncapi: asyncApiSchema.optional(),
        global: z.lazy(() => globalSchema).optional(),
    }),
    baseMenuItemSchema.extend({
        pages: z.lazy(() => pagesSchema),
        openapi: openApiSchema.optional(),
        asyncapi: asyncApiSchema.optional(),
        global: z.lazy(() => globalSchema).optional(),
    }),
    baseMenuItemSchema.extend({
        openapi: openApiSchema.optional(),
        asyncapi: asyncApiSchema.optional(),
        global: z.lazy(() => globalSchema).optional(),
    }),
]);
export const decoratedMenuItemSchema = baseMenuItemSchema.and(z.union([
    z.object({ href: hrefSchema }),
    z.lazy(() => z.object({ languages: decoratedLanguagesSchema, global: globalSchema.optional() })),
    z.lazy(() => z.object({ versions: decoratedVersionsSchema, global: globalSchema.optional() })),
    z.lazy(() => z.object({ dropdowns: decoratedDropdownsSchema, global: globalSchema.optional() })),
    z.lazy(() => z.object({ anchors: decoratedAnchorsSchema, global: globalSchema.optional() })),
    z.lazy(() => z.object({ products: decoratedProductsSchema, global: globalSchema.optional() })),
    z.lazy(() => z.object({ tabs: decoratedTabsSchema, global: globalSchema.optional() })),
    z.lazy(() => z.object({ groups: decoratedGroupsSchema, global: globalSchema.optional() })),
    z.lazy(() => z.object({ pages: decoratedPagesSchema, global: globalSchema.optional() })),
]));
export const menuSchema = z.array(menuItemSchema).describe('Organizing by menu');
export const decoratedMenuSchema = z
    .array(decoratedMenuItemSchema)
    .min(1, 'At least one menu item must be specified')
    .describe('Organizing by menu');
